### ISSUE TITLE
Object tree shaking removes properties that are being used

### ISSUE THREAD

**Issue by voliva** [rollup/rollup#5735]
### Rollup Version

4.27.2

### Operating System (or Browser)

MacOS 14.3

### Node Version (if applicable)

21.6.0

### Link To Reproduction

https://rollupjs.org/repl/?version=4.27.2&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQSUyMiUyMiUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwY29uc3QlMjBnZXRCdWlsZGVyJTIwJTNEJTIwKCklMjAlM0QlM0UlMjAlN0IlNUNuJTIwJTIwY29uc3QlMjB2YXJpYW50U2hhcGVJZCUyMCUzRCUyMCU3QiU1Q24lMjAlMjAlMjAlMjBlcnJvcnMlM0ElMjAxbiUyQyU1Q24lMjAlMjAlMjAlMjBldmVudHMlM0ElMjAybiUyQyU1Q24lMjAlMjAlMjAlMjBjYWxscyUzQSUyMDNuJTJDJTVDbiUyMCUyMCU3RCUzQiU1Q24lNUNuJTIwJTIwY29uc3QlMjBidWlsZFZhcmlhbnQlMjAlM0QlMjAodmFyaWFudFR5cGUpJTIwJTNEJTNFJTIwKCklMjAlM0QlM0UlMjAlN0IlNUNuJTIwJTIwJTIwJTIwdHJ5JTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMHJldHVybiUyMHZhcmlhbnRTaGFwZUlkJTVCdmFyaWFudFR5cGUlNUQlM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTIwY2F0Y2glMjAoXyklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwcmV0dXJuJTIwbnVsbCUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTIwJTIwJTdEJTNCJTVDbiU1Q24lMjAlMjByZXR1cm4lMjAlN0IlNUNuJTIwJTIwJTIwJTIwYnVpbGRDYWxsJTNBJTIwYnVpbGRWYXJpYW50KCU1QyUyMmNhbGxzJTVDJTIyKSUyQyU1Q24lMjAlMjAlMjAlMjBidWlsZEV2ZW50JTNBJTIwYnVpbGRWYXJpYW50KCU1QyUyMmV2ZW50cyU1QyUyMiklMkMlNUNuJTIwJTIwJTIwJTIwYnVpbGRFcnJvciUzQSUyMGJ1aWxkVmFyaWFudCglNUMlMjJlcnJvcnMlNUMlMjIpJTJDJTVDbiUyMCUyMCU3RCUzQiU1Q24lN0QlM0IlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlN0QlN0Q=

```js
export const getBuilder = () => {
  const variantShapeId = {
    errors: 1n,
    events: 2n,
    calls: 3n,
  };

  const buildVariant = (variantType) => () => {
    try {
      return variantShapeId[variantType];
    } catch (_) {
      return null;
    }
  };

  return {
    buildCall: buildVariant("calls"),
    buildEvent: buildVariant("events"),
    buildError: buildVariant("errors"),
  };
};
```

### Expected Behaviour

The object `variantShapeId` should be left as-is, since it's being indexed by an external parameter.

### Actual Behaviour

The object tree-shaking removes properties from `variantShapeId` that should be accessible.

In this specific example, this only happens when wrapping the statement that accesses the object with a try-catch.

### Additional context

This is a regression introduced in 4.27, related to #5420 

The example is the minimum reproduction case based on [existing code of our library](https://github.com/polkadot-api/polkadot-api/blob/5888a34218a6f117f25318791a7b288a508dc392/packages/metadata-builders/src/checksum-builder.ts#L446).

CC @carlosala @lukastaegert @TrickyPi

**Comment by TrickyPi** [comment]
Oh, Rollup will by default cache the first tree-shaking result inside try-statements. One workaround, set `treeshake.tryCatchDeoptimization` to false https://rollupjs.org/configuration-options/#treeshake-trycatchdeoptimization

**Comment by lukastaegert** [comment]
The problem is that polyfills rely on this behavior as some do feature detection by expecting that "valid" code throws.

**Comment by TrickyPi** [comment]
Oh, I get it!

**Comment by TrickyPi** [comment]
Perhaps we could consider limiting this option to only affect `hasEffects` detection and remove its impact on `includePath`?

**Comment by JohnAtFenestra** [comment]
We are noticing this type of problem in a very large project with 4.27.3. Causing crashes in our Quasar/Vue application - don't have a replication for you at the moment, but we've traced that an array is being passed in and comes out the other side as an empty array. Note this has been working for months. Had another serious breakage on Friday that was similar. I can give more details, but right now I need to find a way to force Quasar's build routine to use an older version so we can keep working.

**Comment by lukastaegert** [comment]
Are you sure it is 4.27.3? This version reverts the feature and is mostly equivalent to 4.26

**Comment by JohnAtFenestra** [comment]
Yes. We are sure. Trying some different things now to see if we can narrow it down.

**Comment by JohnAtFenestra** [comment]
A bit of a false alarm, sorry. However, we've clearly traced a bunch of problems to the Vue update on Friday. It just happens that, whatever additionally broke there with the update from Vue 3.3.8 to 3.5.x also creating similar evidencing.
